iocage/lib/NetworkInterface.py

```diff
@@ -152,29 +185,123 @@ def __apply_addresses(
 
         family = "inet6" if ipv6 else "inet"
         for i, address in enumerate(addresses):
+            name = self.current_nic_name
             if (ipv6 is False) and (address.lower() == "dhcp"):
-                command = [self.dhclient_command, self.name]
+                command = [self.dhclient_command, name]
             else:
-                command = [self.ifconfig_command, self.name, family, address]
+                command = [self.ifconfig_command, name, family, address]
 
             if i > 0:
                 command.append("alias")
 
-            self.__exec(command)
+            self._exec(command)
 
             if (ipv6 is True) and (address.lower() == "accept_rtadv"):
-                self.__exec([self.rtsold_command, self.name])
+                self._exec([
+                    self.rtsold_command,
+                    self.current_nic_name
+                ])
 
-    def __exec(
+    def _exec(
         self,
-        command: typing.List[str],
-        force_local: typing.Optional[bool]=False
-    ) -> None:
+        command: typing.List[str]
+    ) -> str:
 
         if self.jail is not None:
-            _, stdout, _ = self.jail.exec(command)
+            stdout, _, _ = self.jail.exec(command)
         else:
-            _, stdout, _ = iocage.lib.helpers.exec(command, logger=self.logger)
+            stdout, _, _ = iocage.lib.helpers.exec(command, logger=self.logger)
+
+        self._handle_exec_stdout(stdout)
 
+        return str(stdout)
+
+    def _handle_exec_stdout(self, stdout: str) -> None:
         if (self.create or self.rename) is True:
             self.name = stdout.strip()
+
+
+class QueuingNetworkInterface(
```

Comment:
This branch changes the initialization process of jails. In the `exec.prestart` hook the host interfaces will be created and firewalled, so that during `exec.start` all interfaces are available in the jail. Everything that is required to prepare the host for a jail (that used to call individual commands) is now squashed into a single file that is executed as `exec.prestart`. The first step of the start script btw. will configure the network interfaces and routes within the jail.

After a jail is started (when it did not already stop at this point) the `exec.poststart` hook will rename the host devices according to the jails JID. This value was unknown before starting the jail, so that we have to work with temporary `ioc:{epair_id}:*` device names. This complicates the shutdown a little bit, but comes with a huge performance gain.